### PR TITLE
update facebook API to version 2.9 and bug fix

### LINKED ISSUE #1247: Facebook login is broken for apps on API version 2.6
1. Connect Zotonic to a Facebook version 2.6 app.
2. Try to log in but don’t give e-mail permission.
3. Try to log in again and you’ll see the screen below. The ‘change permissions’ doesn’t work.

screen shot 2016-04-19 at 15 52 24

### PR DESCRIPTION
**Pull request by andrikurniawan**
### Description

Fix #1247

Update Facebook API to version 2.9 and resolve "do not give email permission" problem

### Checklist

- [ ] documentation updated
- [ ] tests added
- [ ] no BC breaks

### CODE REVIEW

File: modules/mod_facebook/controllers/controller_facebook_redirect.erl [hunk 1]
```diff
@@ -133,9 +133,28 @@ decode_access_token("application/json"++_, Payload) ->
 decode_access_token(_ContentType, Payload) ->
     mochiweb_util:parse_qs(Payload).
 
+% Given user id
+fetch_user_id(AccessToken) ->
+    FacebookUrl = "https://graph.facebook.com/v2.9/me?access_token=" ++ z_utils:url_encode(AccessToken),
```
Comment (ddeboer) [inline modules/mod_facebook/controllers/controller_facebook_redirect.erl]: Please use `z_url:url_encode/1` instead.

File: modules/mod_facebook/controllers/controller_facebook_redirect.erl [hunk 2]
```diff
@@ -133,9 +133,28 @@ decode_access_token("application/json"++_, Payload) ->
...
+    FacebookUrl = "https://graph.facebook.com/v2.9/me?access_token=" ++ z_url:url_encode(AccessToken),
+    case httpc:request(FacebookUrl) of
+        {ok, {{_, 200, _}, _Headers, Payload}} ->
+            {struct, Props} = mochijson:binary_decode(Payload),
+            case proplists:get_value(<<"id">>, Props) of
+                undefined ->
+                    lager:error("[facebook] missing param id ~p", [Props]),
+                    {error, {http_error, FacebookUrl}};
+                _id ->
+                    {ok, jiffy:decode(_id)}
```
Comment (ddeboer) [inline modules/mod_facebook/controllers/controller_facebook_redirect.erl]: This looks like a double decode: on line 141 we already have `mochijson:binary_decode/1`. Is this intentional?

File: modules/mod_facebook/controllers/controller_facebook_redirect.erl [hunk 3]
```diff
@@ -133,9 +133,28 @@ decode_access_token("application/json"++_, Payload) ->
...
+            end;
+        Other ->
+            lager:error("[facebook] error fetching user id [token ~p] ~p", [AccessToken, Other]),
+            {error, {http_error, FacebookUrl, Other}}
+    end.
+
 % Given the access token, fetch data about the user
 fetch_user_data(AccessToken) ->
-    FacebookUrl = "https://graph.facebook.com/v2.3/me?access_token=" ++ z_url:url_encode(AccessToken),
+    {ok, User_id} = fetch_user_id(AccessToken),
```
Comment (arjan) [inline modules/mod_facebook/controllers/controller_facebook_redirect.erl]: Fetching the user id using a separate request seems a bit contrived to me.

The 2.9 API still has the `/me` access point, but in order to get the id you need to specify it in the fields, e.g. `/me?fields=id,name,email`; that way, you only have to do a single request and you can remove the whole `fetch_user_id` function.

Comment (andrikurniawan) [inline modules/mod_facebook/controllers/controller_facebook_redirect.erl]: Thanks Arjan, as I recall, I tried using /me but did not get the email so I decided to add function fetch_user_id

**Comment by mworrell** [comment]
@arjan can you have another look?

**Comment by mworrell** [comment]
Thanks!

I manually merged your changes and added some more in 8d5e458

**Pull request closed without merging**